### PR TITLE
Role and login reporting

### PR DESCRIPTION
**Pull request by jsugarman**
Add a CSV report that extracts data to identify user behaviour for user research. Includes creation and authorisation of a general admin interface for accessing/downloading the report.

see [JIRA Ticket](https://dsdmoj.atlassian.net/browse/PFT-243)

 - [X] add query object
 - [X] add csv publisher
 - [X] add interface for downloading report
 - [X] add superadmin authorization

### CODE REVIEW

**Comment by cleargif** [comment]
@jsugarman I am going to make a couple of tweaks to the views - nothing big

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -81,7 +81,7 @@ def page_title
   end
 
   def render_search_box?
-    logged_in? && !@login_screen && !@editing_mode
+    logged_in? && !@login_screen && !@editing_mode && !@admin_management
```
Comment (jsugarman) [inline app/helpers/application_helper.rb]: change to @admin_management_page for clarity

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -95,6 +97,10 @@ def user_not_authorized
       session[:desired_path] = request.fullpath
       session[:unauthorised_login] = true
       redirect_to new_sessions_path
+    elsif logged_in_regular?
+      warning :unauthorised
+      session[:desired_path] = request.fullpath
```
Comment (jsugarman) [inline app/controllers/application_controller.rb]: is this redundant?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -73,8 +73,10 @@ def desired_path person
 
   def i18n_flash(type, *partial_key, **options)
     full_key = [
-      :controllers, controller_path, *partial_key
+      :controllers, controller_path.tr('/', '.'), *partial_key
     ].join('.')
+    ap full_key
```
Comment (jsugarman) [inline app/controllers/application_controller.rb]: remove

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -73,8 +73,10 @@ def desired_path person
...
+    ap options
```
Comment (jsugarman) [inline app/controllers/application_controller.rb]: remove

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (jsugarman) [inline app/controllers/application_controller.rb]: required?

File: app/queries/user_behavior_query.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+  def joins_sql
+    <<~SQL
+    LEFT JOIN memberships ON memberships.person_id = people.id
+    LEFT JOIN groups ON groups.id = memberships.group_id
+    SQL
+  end
+
+  # rubocop:disable MethodLength
+  def select_ancestors
```
Comment (colinbruce) [inline app/queries/user_behavior_query.rb]: Personal preference but, if you're going to disable MethodLength anyway, I'd stick this into a single massive SQL block.

It'll be easier to parse than jumping up and down to different methods.  YMMV

File: app/queries/user_behavior_query.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        updates_count: rec.updates_count
+      }
+    end
+  end
+
+  private
+
+  def joins_sql
+    <<~SQL
+    LEFT JOIN memberships ON memberships.person_id = people.id
```
Comment (jsugarman) [inline app/queries/user_behavior_query.rb]: indent

File: app/queries/user_behavior_query.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    SQL
+  end
+
+  # rubocop:disable MethodLength
+  def select_ancestors
+    <<~SQL
+      CASE
+        WHEN groups.ancestry_depth IS NOT NULL then
+          (
+            SELECT array_agg(name) AS names
```
Comment (jsugarman) [inline app/queries/user_behavior_query.rb]: extract sub query to one method

**Pull request merged**